scrunch/order.py

```diff
@@ -0,0 +1,576 @@
+            self.name, position=position, before=before, after=after
+        )
+
+
+class Order(object):
+
+    def __init__(self, resource):
+        self.resource = resource
+        self._hier = None
+        self._graph = None
+        self._sync = True
+
+    def _load_hier(self):
+        """
+        subclasses need to overwrite this method in order to get the updated
+        order hierarchy
+        """
+        return self._hier
+
+    def _load_graph(self):
+        self._graph = Group({'__root__': self.hier.graph}, order=self)
+        return self._graph
+
+    @property
+    def hier(self):
+        if self._hier is None:
+            self._load_hier()
+        return self._hier
+
+    @hier.setter
+    def hier(self, _):
+        raise TypeError('Unsupported assignment operation')
+
+    @property
+    def graph(self):
+        if self._graph is None:
+            self._load_graph()
+        return self._graph
+
+    @graph.setter
+    def graph(self, _):
+        raise TypeError('Unsupported assignment operation')
+
+    def _build_graph_structure(self):
+
+        def _get(group):
+            _elements = []
+            for obj in group.elements.values():
+                if isinstance(obj, Group):
+                    _elements.append({
+                        obj.name: _get(obj)
+                    })
+                else:
+                    if isinstance(obj, scrunch.datasets.Variable):
+                        _id = obj.id
+                    else:
+                        _id = obj['id']
+                    _elements.append('../%s/' % _id)
+            return _elements
+
+        return _get(self.graph)
+
+    def update(self):
+        updated_order = {
+            'element': 'shoji:order',
+            'graph': self._build_graph_structure()
+        }
+        try:
+            self.hier.put(updated_order)
+        except (pycrunch.ClientError, pycrunch.ServerError) as e:
+            # Our update to the Hierarchical Order failed. Better reload.
+            self._hier = self._vars = self._graph = None
+            self._load_graph()
+            raise OrderUpdateError(str(e))
+
+    # Proxy methods for the __root__ Group
+
+    def __str__(self):
+        return str(self.graph)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __iter__(self):
+        return self.graph.itervalues()
+
+    def itervalues(self):
+        return self.graph.itervalues()
+
+    def iterkeys(self):
+        return self.graph.iterkeys()
+
+    def keys(self):
+        return self.graph.keys()
+
+    def values(self):
+        return self.graph.values()
+
+    def items(self):
+        return self.graph.items()
+
+    def __getitem__(self, item):
+        return self.graph[item]
+
+
+class DatasetVariablesOrder(Order):
+    def __init__(self, resource):
+        super(DatasetVariablesOrder, self).__init__(resource)
+        self._vars = None
+        self._revision = None
+
+    def _load_hier(self):
```

Comment:
See comment above, rename `_hier` to store the shoji:order as `self.order` straight, no need for this lazy loading pattern